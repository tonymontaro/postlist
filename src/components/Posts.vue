<template>
  <section id="posts">
    <ul>
      <li v-for="post in posts" :key="post.id">
        {{ post.id }}. {{ post.title }}
      </li>
    </ul>
  </section>
</template>

<script>
export default {
  name: "posts",
  data() {
    return {
      posts: this.$store.state.posts.posts
    };
  },
  mounted() {
    this.getPosts();
  },
  methods: {
    async getPosts() {
      if (this.posts.length == 0) {
        await this.$store.dispatch("posts/getPosts");
      }
    }
  }
};
</script>

<style scoped>
li {
  display: block;
  padding: 20px;
  border: 1px solid gray;
  margin: 5px 0;
  width: 500px;
}
</style>
