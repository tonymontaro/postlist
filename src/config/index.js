export const POST_URL = "https://jsonplaceholder.typicode.com/posts";
export const POST_LIMIT = 5;
export const BOX_STYLE = [
  "p-3",
  "py-5",
  "bg-white",
  "shadow-lg",
  "rounded-sm",
  "white-box",
  "w-full"
];
