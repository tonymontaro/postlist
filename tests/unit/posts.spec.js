import axios from "axios";
import Vuex from "vuex";
import { createLocalVue, mount, shallowMount } from "@vue/test-utils";

import Posts from "@/components/Posts";
import postStore from "@/store/posts";
import { postUrl } from "@/config";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("Component", () => {
  test("renders post items", () => {
    const store = new Vuex.Store({
      state: {
        posts: {
          posts: [
            { id: 1, title: "1" },
            { id: 2, title: "2" }
          ]
        }
      },
      actions: { "posts/getPosts": jest.fn() }
    });

    const wrapper = mount(Posts, {
      localVue,
      store
    });
    expect(wrapper.findAll('span[name="slide"]').length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("Methods", () => {
  let store;
  let actions;
  let state;

  beforeEach(() => {
    state = { posts: { posts: [] } };
    actions = { "posts/getPosts": jest.fn() };
    store = new Vuex.Store({
      state,
      actions
    });
  });

  test("getPosts calls the getPosts action", () => {
    const expected = [{ id: 1, title: "Title" }];
    axios.get.mockResolvedValue({ data: expected });
    const wrapper = mount(Posts, {
      localVue,
      store
    });

    expect(actions["posts/getPosts"]).toHaveBeenCalledTimes(1);
  });
});

describe("Store", () => {
  test("store loads successfully", () => {
    const store = new Vuex.Store({
      state: { posts: { posts: [] } },
      actions: { "posts/getPosts": jest.fn() }
    });
    const wrapper = mount(Posts, {
      localVue,
      store
    });
    expect(wrapper.vm.posts).toHaveLength(0);
  });

  test("action.getPosts makes api get request", async () => {
    const { getPosts } = postStore.actions;
    const expected = [{ id: 1, title: "Title" }];
    axios.get.mockResolvedValue({ data: expected });

    const res = await getPosts({ commit: () => {} });

    expect(res).toEqual(expected);
    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(axios.get).toHaveBeenCalledWith(postUrl);
  });

  test("mutations.getPostsSuccess sets the state to first five items", () => {
    const { getPostsSuccess } = postStore.mutations;
    const vuexState = { posts: [] };
    const posts = [1, 2, 3, 4, 5, 6, 7];
    const expected = [1, 2, 3, 4, 5];

    getPostsSuccess(vuexState, posts);
    expect(vuexState.posts).toEqual(expected);
  });
});
