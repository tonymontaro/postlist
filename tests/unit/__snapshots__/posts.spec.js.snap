// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders post items 1`] = `
<section id="posts">
  <h1 class="text-xl text-white mb-5">Sortable Post List</h1>
  <ul><span name="slide"><post-stub movedown="function () { [native code] }" moveup="function () { [native code] }" post="[object Object]" index="0"></post-stub><post-stub movedown="function () { [native code] }" moveup="function () { [native code] }" post="[object Object]" index="1"></post-stub><post-stub movedown="function () { [native code] }" moveup="function () { [native code] }" post="[object Object]" index="2"></post-stub><post-stub islastpost="true" movedown="function () { [native code] }" moveup="function () { [native code] }" post="[object Object]" index="3"></post-stub></span></ul>
</section>
`;
